app/controllers/api/openid_connect/authorizations_controller.rb

```diff
@@ -0,0 +1,249 @@
+        handle_params_error(error, "The request was malformed: please double check the client id and redirect uri.")
+      end
+
+      rescue_from OpenSSL::SSL::SSLError do |e|
+        logger.info e.backtrace[0, 10].join("\n")
+        handle_params_error("bad_request", e.message)
+      end
+
+      rescue_from JSON::JWS::VerificationFailed do |e|
+        logger.info e.backtrace[0, 10].join("\n")
+        handle_params_error("bad_request", e.message)
+      end
+
+      before_action :auth_user_unless_prompt_none!
+
+      def new
+        auth = Api::OpenidConnect::Authorization.find_by_client_id_and_user(params[:client_id], current_user)
+        reset_auth(auth)
+        if logged_in_before?(params[:max_age])
+          reauthenticate(params)
+        elsif params[:prompt]
+          prompt = params[:prompt].split(" ")
+          handle_prompt(prompt, auth)
+        else
+          handle_authorization_form(auth)
+        end
+      end
+
+      def create
+        restore_request_parameters
+        process_authorization_consent(params[:approve])
+      end
+
+      def destroy
+        authorization = Api::OpenidConnect::Authorization.find_by(id: params[:id])
+        if authorization
+          authorization.destroy
+        else
+          flash[:error] = I18n.t("api.openid_connect.authorizations.destroy.fail", id: params[:id])
+        end
+        redirect_to api_openid_connect_user_applications_url
+      end
+
+      private
+
+      def reset_auth(auth)
+        return unless auth
+        auth.o_auth_access_tokens.destroy_all
+        auth.id_tokens.destroy_all
+        auth.code_used = false
+        auth.save
+      end
+
+      def handle_prompt(prompt, auth)
+        if prompt.include? "select_account"
+          handle_params_error("account_selection_required",
+                              "There is no support for choosing among multiple accounts")
+        elsif prompt.include?("login")
+          new_params = params.merge!(prompt: params[:prompt].remove("login"))
+          reauthenticate(new_params)
+        elsif prompt.include? "consent"
+          request_authorization_consent_form
+        else
+          handle_authorization_form(auth)
+        end
+      end
+
+      def reauthenticate(params)
+        sign_out current_user
+        redirect_to new_api_openid_connect_authorization_path(params)
+      end
+
+      def handle_authorization_form(auth)
+        if auth
+          process_authorization_consent("true")
+        else
+          request_authorization_consent_form
+        end
+      end
+
+      def request_authorization_consent_form
+        add_claims_to_scopes
+        endpoint = Api::OpenidConnect::AuthorizationPoint::EndpointStartPoint.new(current_user)
+        handle_start_point_response(endpoint)
+      end
+
+      def add_claims_to_scopes
+        return unless params[:claims]
+        claims_json = JSON.parse(params[:claims])
+        return unless claims_json
```

Comment:
And I strongly dislike blocks of assignments and returns. But oh well, let's keep it that way then. :)